tasks/message/message.go

```diff
@@ -0,0 +1,246 @@
+package message
+
+import (
+	"context"
+	"time"
+
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"go.opentelemetry.io/otel/api/global"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/sentinel-visor/lens"
+	messagemodel "github.com/filecoin-project/sentinel-visor/model/messages"
+	"github.com/filecoin-project/sentinel-visor/model/visor"
+	"github.com/filecoin-project/sentinel-visor/storage"
+	"github.com/filecoin-project/sentinel-visor/wait"
+)
+
+const idleSleepInterval = 60 * time.Second   // time to wait if the processor runs out of blocks to process
+const batchInterval = 100 * time.Millisecond // time to wait between batches
+
+var log = logging.Logger("message")
+
+var timeNow = time.Now
+
+func NewMessageProcessor(d *storage.Database, node lens.API, leaseLength time.Duration, batchSize int, maxHeight int64) *MessageProcessor {
+	return &MessageProcessor{
+		node:        node,
+		storage:     d,
+		leaseLength: leaseLength,
+		batchSize:   batchSize,
+		maxHeight:   maxHeight,
+	}
+}
+
+// MessageProcessor is a task that processes blocks to detect messages and persists
+// their details to the database.
+type MessageProcessor struct {
+	node        lens.API
+	storage     *storage.Database
+	leaseLength time.Duration // length of time to lease work for
+	batchSize   int           // number of tipsets to lease in a batch
+	maxHeight   int64         // limit processing to tipsets equal to or below this height
+}
+
+// Run starts processing batches of tipsets and blocks until the context is done or
+// an error occurs.
+func (p *MessageProcessor) Run(ctx context.Context) error {
+	// Loop until context is done or processing encounters a fatal error
+	return wait.RepeatUntil(ctx, batchInterval, p.processBatch)
```

Comment:
The for loop inside `wait.RepeatUntil` interfers with the schedulers `Run` loop. With `wait.RepeatUntil` the 4 message processors will only run once and then they block on something. Still digging...

Reply:
Removing `wait.RepeatUntil` allows for the processors to keep restarting as intended but without the sleep interval between each run of the task.

Reply:
This is the deadlock I mentioned. Should be fixed in current branch

Reply:
it is 👍